pubnub-core/src/subscription/subscribe_loop.rs

```diff
@@ -0,0 +1,282 @@
+                    // Termination requested, break the loop.
+                    break;
+                }
+
+                // Control signalled we can continue with the polling, however
+                // we literally need to `continue` here in order to force rerun
+                // the loop from the beginning.
+                // We rely on the in-flight request to be properly cleaned up,
+                // since their futures are being dropped here.
+                continue;
+            }
+            Either::Right((res, _)) => {
+                match res {
+                    Ok(v) => v,
+                    Err(err) => {
+                        // TODO: add some kind of circut breaker.
+                        // Report error and retry - maybe it'd work this time.
+                        error!("Transport error while polling: {:?}", err);
+                        continue;
+                    }
+                }
+            }
+        };
+
+        // Send ready message when the subscribe loop is capable of receiving
+        // messages.
+        // This is intended to signal the readiness (and the healthiness) of
+        // the setup. It is invoked after the `Ok` result from the request
+        // future, guaranteing that Transport was able to perform successfully
+        // at least once.
+        if let Some(ready_tx) = ready_tx.take() {
+            if let Err(err) = ready_tx.send(()) {
+                error!("Error sending ready message: {:?}", err);
+                break;
+            }
+        }
+
+        // Save Timetoken for next request
+        timetoken = next_timetoken;
+
+        debug!("messages: {:?}", messages);
+        debug!("timetoken: {:?}", timetoken);
+
+        // Distribute messages to each listener.
+        for message in messages {
+            let route = message
+                .route
+                .clone()
+                .unwrap_or_else(|| message.channel.clone());
+            debug!("route: {}", route);
+
+            // TODO: provide a better interface and remove the potentially
+            // unsound `get` and `get_mut` from the registry API.
+            let listeners = channels.get_mut(&route).unwrap();
+
+            debug!("Delivering to {} listeners...", listeners.len());
+            for channel_tx in listeners.iter_mut() {
+                if let Err(error) = channel_tx.send(message.clone()).await {
+                    error!("Delivery error: {:?}", error);
+                }
+            }
+        }
+    }
+
+    debug!("Stopping subscribe loop");
+
+    if let Some(ref mut exit_tx) = exit_tx {
+        exit_tx.send(()).await.expect("Unable to send exit message");
+    }
+}
+
+/// Encodes action to be taken in response to control command.
+#[derive(Debug)]
+enum ControlOutcome {
+    Terminate,
+    CanContinue,
+}
+
+/// Handle a control command.
+///
+/// This is split out from the `select!` macro used in the subscribe loop.
```

Comment:
The caller is no longer using `select!` :+1:

Reply:
Yep, I have to clean this up.